firebase-firestore/src/main/java/com/google/firebase/firestore/model/Document.java

```diff
@@ -36,59 +40,107 @@
   }
 
   private static final Comparator<Document> KEY_COMPARATOR =
-      new Comparator<Document>() {
-        @Override
-        public int compare(Document left, Document right) {
-          return left.getKey().compareTo(right.getKey());
-        }
-      };
+      (left, right) -> left.getKey().compareTo(right.getKey());
 
   /** A document comparator that returns document by key and key only. */
   public static Comparator<Document> keyComparator() {
     return KEY_COMPARATOR;
   }
 
-  private final ObjectValue data;
-
-  private final DocumentState documentState;
+  /** Creates a new Document from an existing ObjectValue. */
+  public static Document fromObjectValue(
+      DocumentKey key, SnapshotVersion version, ObjectValue data, DocumentState documentState) {
+    return new Document(key, version, documentState) {
+      @Nullable
+      @Override
+      public com.google.firestore.v1.Document getProto() {
+        return null;
+      }
+
+      @Override
+      public ObjectValue getData() {
+        return data;
+      }
+
+      @Nullable
+      @Override
+      public FieldValue getField(FieldPath path) {
+        return data.get(path);
+      }
+    };
+  }
 
   /**
-   * Memoized serialized form of the document for optimization purposes (avoids repeated
-   * serialization). Might be null.
+   * Creates a new Document from a Proto representation.
+   *
+   * <p>The Proto is only converted to an ObjectValue if the consumer calls `getData()`.
    */
-  private final com.google.firestore.v1.Document proto;
-
-  public @Nullable com.google.firestore.v1.Document getProto() {
-    return proto;
+  public static Document fromProto(
+      RemoteSerializer serializer,
+      com.google.firestore.v1.Document proto,
+      DocumentState documentState) {
+    DocumentKey key = serializer.decodeKey(proto.getName());
+    SnapshotVersion version = serializer.decodeVersion(proto.getUpdateTime());
+
+    hardAssert(
+        !version.equals(SnapshotVersion.NONE), "Found a document Proto with no snapshot version");
+
+    return new Document(key, version, documentState) {
```

Comment:
This kind of anonymous class makes for some very difficult to understand stack traces when they're reported to us.

A better arrangement is to make named private static classes, e.g. `EagerDocument` and `LazyProtoDocument` that we explicitly instantiate.